src/core/TESTING.md

```diff
@@ -0,0 +1,210 @@
+
+```tsx
+/** public/plugin.ts */
+class Plugin {
+  setup(core) {
+    core.application.register({
+      // id, title, etc.
+      async mount(params) {
+        const [{ renderApp }, [coreStart, startDeps]] = await Promise.all([
+          import('./application'),
+          core.getStartServices()
+        ]);
+
+        return renderApp(params, coreStart, startDeps);
+      }
+    })
+  }
+}
+```
+
+We _could_ still write tests for this logic:
+
+```ts
+/** public/plugin.test.ts */
+jest.mock('./application', () => ({ renderApp: jest.fn() }));
+import { coreMock } from 'src/core/public/mocks';
+import { renderApp: renderAppMock } from './application';
+import { Plugin } from './plugin';
+
+describe('Plugin', () => {
+  it('registers an app', () => {
+    const coreSetup = coreMock.createSetup();
+    new Plugin(coreMock.createPluginInitializerContext()).setup(coreSetup);
+    expect(coreSetup.application.register).toHaveBeenCalledWith({
+      id: 'myApp',
+      mount: expect.any(Function)
+    });
+  });
+
+  // Test the glue code from Plugin -> renderApp
+  it('application.mount wires up dependencies to renderApp', async () => {
+    const coreSetup = coreMock.createSetup();
+    const [coreStartMock, startDepsMock] = await coreSetup.getStartServices();
+    const unmountMock = jest.fn();
+    renderAppMock.mockReturnValue(unmountMock);
+    const params = { element: document.createElement('div'), appBasePath: '/fake/base/path' };
+
+    new Plugin(coreMock.createPluginInitializerContext()).setup(coreSetup);
+    // Grab registered mount function
+    const mount = coreSetup.application.register.mock.calls[0][0].mount;
+
+    const unmount = await mount(params);
+    expect(renderAppMock).toHaveBeenCalledWith(params, coreStartMock, startDepsMock);
+    expect(unmount).toBe(unmountMock);
+  });
+});
+```
+
+But you may find that you're just asserting the same things that would be covered by type-checks. The more interesting logic is in `renderApp`:
+
+```ts
+/** public/application.ts */
+import React from 'react';
+import ReactDOM from 'react-dom';
+
+import { AppMountParams, CoreStart } from 'src/core/public';
+import { AppRoot } from './components/app_root';
+
+export const renderApp = ({ element, appBasePath }: AppMountParams, core: CoreStart, plugins: MyPluginDepsStart) => {
+  // uiSettings subscription
+  const uiSettingsClient = core.uiSettings.client;
+  const settings$ = combineLatest(uiSettingClient.get$('mysetting1'), uiSettingClient.get$('mysetting2'));
+  ReactDOM.render(
+    <AppRoot routerBasePath={appBasePath} core={core} plugins={plugins} settings$={settings$} />,
+    element
+  );
+
+  return () => {
+    // Unmount UI
+    ReactDOM.unmountComponentAtNode(element);
+    // Close any subscriptions
+    settings$.complete();
```

Comment:
Well. If a plugin **doesn't** subscribe manually, it **needn't** to unsubscribe. We must explain this in the manual.
Given your example.
`settings$` is Observable. A plugin cannot complete it. It will be completed if the unsubscription is implemented in the `useObservable` hook correctly.
The better example could be if a plugin implements polling for a every new settings$ value and implements a subscription **manually**.
```ts
describe('renderApp', () => {
  it('mounts and unmounts UI', () => {
    const params = { element: document.createElement('div'), appBasePath: '/fake/base/path' };
    const core = coreMock.createStart();
    const settings$ = new Rx.Subject<any>();
    core.uiSettings.get$.mockReturnValue(settings$);

    const unmount = renderApp(params, core);
    // Verify some expected DOM element is rendered into the element
    expect(params.element.querySelector('.some-app-class')).not.toBeUndefined();
    // Verify that subscribed to settings$
    expect(settings$.observers.length).toBeGreaterThan(0);

    unmount();
    // Verify the element is empty after unmounting
    expect(params.element.innerHTML).toEqual('');
    // Verify that unsubscribed correctly
    expect(settings$.observers.length).toBe(0);
  });
});
```

We can extract some of this logic in [a custom matcher](https://jestjs.io/docs/en/expect#custom-matchers-api) to be able to write something like this:
```js
expect(settings$).toHaveNoSubscriptions();
```

Reply:
Thanks for the input here, Subjects make this much easier / possible. I think we should encourage this pattern for more fine-grained results but rely on something like drool to make this foolproof.